tutorials/Tutorial17_GPL.py

```diff
@@ -0,0 +1,249 @@
+# -*- coding: utf-8 -*-
+"""Generative-Pseudo-Labeling-for-Domain-Adaptation-of-Dense-Retrievals.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1Tz9GSzre7JfvXDDKe7sCnO0FMuDViMnN
+
+# Generative Pseudo Labeling for Domain Adaptation of Dense Retrievals
+#### Note: Adapted to Haystack from Nils Riemers' original [notebook](https://colab.research.google.com/gist/jamescalam/d2c888775c87f9882bb7c379a96adbc8/gpl-domain-adaptation.ipynb#scrollTo=183ff7ab)
+
+NLP models we use every day were trained on a large corpus of data, reflecting the World from the past. What if some significant world-changing events occur in the meantime, and we want our models to know about them? Like the COVID pandemic, for example? It doesn't make sense to retrain the models from scratch as that would be wasteful. What if we could update the models with new essential data instead of retraining the models from scratch? **Generative Pseudo Labeling (GPL)** to the rescue.
+
+In the below example, we demonstrate this for a simple query: "How is COVID-19 transmitted?".
+
+As model, we use TAS-B: A DistilBERT model that achieves state-of-the-art performance on MS MARCO (500k queries from Bing Search Engine). Both DistilBERT and MS MARCO were created with data from 2018 and before, hence, it lacks the knowledge of any COVID-related information.
+
+In your example we use a small collection of just 4 documents. If you search with this model, you get the following results (dot-score & document):
+- 94.84	Ebola is transmitted via direct contact with blood
+- 92.87	HIV is transmitted via sex or sharing needles
+- 92.31	Corona is transmitted via the air
+- 91.54	Polio is transmitted via contaminated water or food
+
+
+As we see, the correct document is just ranked on 3rd place behind how Ebola and HIV are transmitted.
+
+## Efficient Domain Adaptation with GPL
+This notebook demonstrates [Generative Pseudo Labeling (GPL)](https://arxiv.org/abs/2112.07577), an efficient approach to adapt existing dense retrieval models to new domains & data.
+
+We get a collection 10k scientific papers on COVID-19 and then fine-tune within 15-60 minutes (depending on your GPU) to include the new COVID knowledge into our model.
+
+If we search again with the updated model, we get the search results we would expect:
+- Query: How is COVID-19 transmitted
+- 97.70	Corona is transmitted via the air
+- 96.71	Ebola is transmitted via direct contact with blood
+- 95.14	Polio is transmitted via contaminated water or food
+- 94.13	HIV is transmitted via sex or sharing needles
+"""
+
+from sentence_transformers import SentenceTransformer, util
+from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+from datasets import load_dataset
+
+
+def tutorial17_gpl():
+    # We load the TAS-B model, a state-of-the-art model trained on MS MARCO
+    max_seq_length = 200
```

Comment:
would it be helpful to add a comment with some more explanations what each code line does or do we expect the users to understand this?